manta-trusted-setup/src/ceremony/signature.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+//! Signature Scheme for trusted setup.
+use crate::ceremony::CeremonyError;
+/// Public Key of participant
+pub trait HasPublicKey {
+    /// Public Key of participant
+    type PublicKey;
+
+    /// Returns the public key of the participant.
+    fn public_key(&self) -> Self::PublicKey;
+}
+
+/// Signature Scheme types
+pub trait SignatureScheme {
+    /// Public Key
+    type PublicKey;
+
+    /// Private Key
+    type PrivateKey;
+
+    /// Domain Tag specific to the use case
+    type DomainTag: ?Sized + 'static;
+
+    /// Signature
+    type Signature;
+}
+
+/// Verifiable Message
+pub trait Verify<S>
+where
+    S: SignatureScheme + ?Sized,
+{
+    /// Signature specific to this message
+    type Signature: Sized;
+    /// Verify the integrity of the message
+    fn verify_integrity(
+        &self,
+        domain_tag: &S::DomainTag,
+        public_key: &S::PublicKey,
+        signature: &Self::Signature,
+    ) -> Result<(), CeremonyError>;
+}
+
+/// Signable Message
+pub trait Sign<S>
+where
+    S: SignatureScheme + ?Sized,
+{
+    /// Signature specific to this message
+    type Signature: Sized;
+    /// Sign the message
+    fn sign(
+        &self,
+        domain_tag: &S::DomainTag,
+        public_key: &S::PublicKey,
+        private_key: &S::PrivateKey,
+    ) -> Result<Self::Signature, CeremonyError>;
```

Comment:
in my understanding, we only need the private key to sign a message. Why should we use both private key and public key here?